### PR TITLE
Icefall-based, Transducer ASR Model

### PR DESCRIPTION
**Pull request by HSTEHSTEHSTE**
This is a preliminary pull request, containing the ART-adaptation for Icefall-based ASR Models

### CODE REVIEW

**Comment by beat-buesser** [comment]
Hi @HSTEHSTEHSTE Thank you for your pull request! Could you please target one of the dev branches, either `dev_1.17.0` or `dev_1.18.0`?

File: art/estimators/speech_recognition/pytorch_icefall.py [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+
+        feature_list = []
+        num_frames = []
+        supervisions = {}
+
+        for i in range(len(x)):
+            isnan = torch.isnan(x[i])
+            nisnan = torch.sum(isnan).item()
+            if nisnan > 0:
+                logging.info("input isnan={}/{} {}".format(nisnan, x[i].shape, x[i][isnan], torch.max(torch.abs(x[i]))))
```
Comment (f4str) [inline art/estimators/speech_recognition/pytorch_icefall.py:257]:
```suggestion
                logging.info(f"input isnan={nisnan}/{x[i][isnan]} {torch.max(torch.abs(x[i]))}")
```
Use f-strings to avoid confusion with string formatting

File: art/estimators/speech_recognition/pytorch_icefall.py [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+            low_freq: float = 20.0
+            high_freq: float = -400.0
+            num_mel_bins: int = 40
+            use_energy: bool = False
+            vtln_low: float = 100.0
+            vtln_high: float = -500.0
+            vtln_warp: float = 1.0
+
+        params = asdict(FbankConfig())
+        params.update({"sample_frequency": 16000, "snip_edges": False, "num_mel_bins": 23})
```
Comment (f4str) [inline art/estimators/speech_recognition/pytorch_icefall.py:245]: The dataclass seems a bit extraneous. What's the issue with just using a dictionary?

File: art/estimators/speech_recognition/pytorch_icefall.py [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+
+        assert len(original_output[0]) == 1
+        num_batch = len(original_output[0])
+        decoded_output = []
+
+        for sample_index in range(num_batch):
+            features, _, _ = self.transform_model_input(x=masked_adv_input[sample_index])
+            x_lens = torch.tensor([features.shape[1]]).to(torch.int32).to(self.device)
+            y = k2.RaggedTensor(original_output[sample_index])
+            loss = self.transducer_model(x=features, x_lens=x_lens, y=y)
```
Comment (f4str) [inline art/estimators/speech_recognition/pytorch_icefall.py:361]: Should the loss not also be returned?

File: .github/workflows/ci-icefall.yml [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+              libgl1-mesa-glx \
+              libx11-xcb1 \
+              git \
+              gcc \
+              mono-mcs \
+              libavcodec-extra \
+              ffmpeg \
+              curl \
+              libsndfile-dev \
+              libsndfile1 \
```
Comment (f4str) [inline .github/workflows/ci-icefall.yml:55]: Is there a reason why this is done as two `apt install` steps? Can't they be combined into one?

File: art/estimators/speech_recognition/pytorch_icefall.py [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+"""
+This module implements the task specific estimator for Icefall, an end-to-end speech recognition toolkit based on
+k2-fsa.
+
+| Repository link: https://github.com/k2-fsa/icefall/tree/master
+"""
+import ast
+from argparse import Namespace
+import logging
+from typing import Dict, List, Optional, Tuple, TYPE_CHECKING, Union
```
Comment (f4str) [inline art/estimators/speech_recognition/pytorch_icefall.py:27]: Make sure to run the style checks and see any issues such as unused imports

File: .github/workflows/ci-icefall.yml [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+          git clone https://github.com/k2-fsa/icefall.git
+          cd icefall
+          pip install -r requirements.txt
+          pip install pytest
+          pip install pytest-cov
+          pip install kaldiio
+          pip install tensorflow
+          cd egs/fluent_speech_commands/SLU/transducer
+          wget -O epoch-6.pt https://www.dropbox.com/scl/fi/97wvdjmbuyj13kpzhricc/epoch-6.pt?rlkey=7mehc4v41fovfz0ksbt98krry&dl=0
+          wget --load-cookies /tmp/cookies.txt "https://docs.google.com/uc?export=download&confirm=$(wget --quiet --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=1K-CaHTtJ-zZ5oW0us0OaKQmsD2-cO0xN' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\1\n/p')&id=1K-CaHTtJ-zZ5oW0us0OaKQmsD2-cO0xN" -O frames.zip && rm -rf /tmp/cookies.txt1K-CaHTtJ-zZ5oW0us0OaKQmsD2-cO0xN
```
Comment (f4str) [inline .github/workflows/ci-icefall.yml:75]: Can we change this to a dropbox instead of. google drive?

File: art/estimators/speech_recognition/pytorch_icefall.py [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+    Icefall.
+
+    | Repository link: https://github.com/k2-fsa/icefall/tree/master
+    """
+
+    from icefall.utils import AttributeDict
+    from pathlib import Path
+    import k2
+
+    estimator_params = PyTorchEstimator.estimator_params + ["icefall_config_filepath"]
```
Comment (f4str) [inline art/estimators/speech_recognition/pytorch_icefall.py:59]:
```suggestion
    estimator_params = PyTorchEstimator.estimator_params
```
The `icefall_config_filepath` parameter is not longer in the constructor.

File: art/estimators/speech_recognition/pytorch_icefall.py [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+            xx = xx.to(self._device)
+            feat_i = torchaudio.compliance.kaldi.fbank(xx.unsqueeze(0), **params)  # [T, C]
+            feat_i = feat_i.transpose(0, 1)  # [C, T]
+            feature_list.append(feat_i)
+            num_frames.append(feat_i.shape[1])
+
+        indices = sorted(range(len(feature_list)), key=lambda i: feature_list[i].shape[1], reverse=True)
+        indices = torch.LongTensor(indices)
+        num_frames = torch.IntTensor([num_frames[idx] for idx in indices])
+        start_frames = torch.zeros(len(x), dtype=torch.int)
```
Comment (f4str) [inline art/estimators/speech_recognition/pytorch_icefall.py:269]:
```suggestion
        indices = torch.tensor(indices, dtype=torch.int64, device=self._device)
        num_frames = torch.tensor([num_frames[idx] for idx in indices], dtype=torch.int32, device=self._device)
        start_frames = torch.zeros(len(x), dtype=torch.int32, device=self._device)
```
Let's use the modern torch tensor instantiation on the proper device for optimal performance.

File: art/estimators/speech_recognition/pytorch_icefall.py [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+    def input_shape(self) -> Tuple[int, ...]:
+        """
+        Return the shape of one input sample.
+
+        :return: Shape of one input sample.
+        """
+        return self._input_shape  # type: ignore
+
+    @property
+    def model(self):
```
Comment (f4str) [inline art/estimators/speech_recognition/pytorch_icefall.py:313]:
```suggestion
    def model(self) -> "torch.nn.Module":
```
Missing return type

File: contrib/lingvo-patched-decoder.py [hunk 10]
```diff
@@ -41,1488 +41,1444 @@
```
Comment (f4str) [inline contrib/lingvo-patched-decoder.py:41]: Undo the changes to this file. It probably should not have been committed.

File: tests/estimators/speech_recognition/test_pytorch_icefall.py [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+    )
+
+    vocab_size = 1
+    try:
+        with open(Path(params.lang_dir) / "lexicon_disambig.txt") as lexicon_file:
+            for line in lexicon_file:
+                if len(line.strip()) > 0:  # and '<UNK>' not in line and '<s>' not in line and '</s>' not in line:
+                    vocab_size += 1
+    except:
+        pass
```
Comment (f4str) [inline tests/estimators/speech_recognition/test_pytorch_icefall.py:184]: The exception is being swallowed. It should be handled hence the CodeQL warning.

File: tests/estimators/speech_recognition/test_pytorch_icefall.py [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+    # 0 is blank
+    id = 1
+    try:
+        with open(Path(params.lang_dir) / "lexicon_disambig.txt") as lexicon_file:
+            for line in lexicon_file:
+                if len(line.strip()) > 0:
+                    word2id[line.split()[0]] = id
+                    id += 1
+    except:
+        pass
```
Comment (f4str) [inline tests/estimators/speech_recognition/test_pytorch_icefall.py:228]: The exception is being swallowed. It should be handled hence the CodeQL warning.

**Review by f4str — CHANGES_REQUESTED** [review]
This looks good, thanks for implementing the new ASR model. I left a few comments on things that should be addressed.

File: art/estimators/speech_recognition/pytorch_icefall.py (hunk 3 above)
Comment (HSTEHSTEHSTE) [reply art/estimators/speech_recognition/pytorch_icefall.py:361]: I initially assumed so too but apprently not (if the other estimator implementations are anything to go by)
